README.md
# BSDE: Barycenter Single-cell RNAseq Differential Expression analysis for case-control studies
<img src="doc/Fig1.png" alt="rstudio" style="zoom:60%;" />


## Installation
```r
# install.packages("devtools")
devtools::install_github("mqzhanglab/BSDE", build_vignettes = TRUE)
```

In case of missing dependency, the following R packages need to be installed from [BioConductor](https://bioconductor.org/):
```R
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("scater")
BiocManager::install("SingleCellExperiment")
```

### Python dependency

Additionally, the package calls (via [reticulate](https://rstudio.github.io/reticulate/)) Python module [POT](https://pythonot.github.io/) to compute optimal transport. To things work properly, POT should be pre-installed in a [conda](https://docs.conda.io/en/latest/) environment named `r-reticulate`. This can be set up automatically the first time BSDE is loaded. 

But if that fails, you can install manually:

1. First, make sure conda/miniconda is installed. You can follow the instructions [here](https://conda.io/projects/conda/en/latest/user-guide/install/index.html).

2. Run the following in a terminal (line by line):

   ```bash
   conda create --name r-reticulate python=3.7
   conda activate r-reticulate
   conda install -c conda-forge pot
   ```

#### Still have problem loading?

If you are using [RStudio](https://www.rstudio.com/), make sure the Python interpreter is not overridden to something else. In the `Preferences` → `Python`, select interpreter from the `r-reticulate` conda environment. Then, restart R and try `library(BSDE)`. You should be able to see message "Python environment loaded".

<img src="doc/rstudio.png" alt="rstudio" style="zoom:30%;" />



## Documents

- Call `vignette("BSDE")` to see a short introduction.
- See [here](https://github.com/mqzhanglab/BSDE_pipeline) for examples on data analysis. 



NEWS.md
## CHANGES IN VERSION 0.10.0

- Pre-release

R/zzz.R
.onLoad <- function(libname, pkgname) {
  tryCatch({
    reticulate::use_condaenv("r-reticulate", required = TRUE)
    reticulate::py_run_file(system.file("op_functions.py", package = "BSDE"))
    packageStartupMessage("Python environment loaded.")
    # enable parallelism
    n.cores <- max(parallel::detectCores() - 2, 1)
    doParallel::registerDoParallel(cores = n.cores)
    packageStartupMessage(sprintf("Running in parallel on %d cores.", n.cores))
  }, error = function(.e) {
    message(.e)
    message("\nPython environment cannot be loaded.\n")
    .install <- utils::askYesNo("Do you want to set up a conda environment for Python dependency?\nBefore you proceed, please make sure conda (or minoconda) is installed.\nSee https://docs.conda.io/en/latest/miniconda.html")
    if (.install) {
      tryCatch({
        reticulate::conda_create("r-reticulate", python_version = 3.7)
        reticulate::conda_install("r-reticulate", c("numpy", "cython"))
        reticulate::conda_install("r-reticulate", "POT")
        message("Environment `r-reticulate' is set up. Try `library(BSDE)' again.")
      }, error = function(.e) {
        stop("Environment cannot be loaded. See https://github.com/mqzhanglab/BSDE for installation instructions.")
      })
    }
  })
}


R/op_functions.R
decimalplaces <- function(x) {
  if ((x %% 1) != 0) {
    nchar(strsplit(sub('0+$', '', as.character(x)), ".", fixed=TRUE)[[1]][[2]])
  } else {
    return(0)
  }
}

#' Calculate the BSDE p-value for a SingleCellExperiment object.
#'
#' @param sce A SingleCellExperiment object.
#'   See \code{\link[SingleCellExperiment]{SingleCellExperiment}}.
#' @param perm_num The number of permutations for computing Monte Carlo p-value.
#' @param unif_round_unit Bin width for empirical histogram (default: 0.2).
#' @param weight Weight vector (non-negative, length=number of subjects) for subjects.
#'   Can be used for inverse probability weighting to control for confounding.
#'   Default: 1 for equal weights.
#' @param shrink If \code{TRUE}, the range with zero count will be removed to
#'   cope with sparsity and speed up the computation. Use \code{TRUE} if
#'   extremely high expressions are present. Default: \code{FALSE}.
#' @return a list of p-values
#' @export
BSDE <- function(sce, perm_num=200, unif_round_unit=0.2, weight=1, shrink=FALSE) {
  sce <- scater::logNormCounts(sce)
  count_log <- SingleCellExperiment::logcounts(sce)
  cur_ind <- SingleCellExperiment::colData(sce)$individual
  cur_condition <- SingleCellExperiment::colData(sce)$condition
  gene_id <- SingleCellExperiment::rowData(sce)$gene_id
  op_pval <- plyr::laply(1:nrow(count_log),
                        function(i_g) {
                          cur_count <- count_log[i_g, ]
                          if(sum(cur_count)==0){
                            return(1)
                          } else {
                          return(tryCatch({
                            BSDE::cal_w2_pval(
                              count_per_gene = cur_count,
                              meta_individual = cur_ind,
                              meta_phenotype = cur_condition,
                              perm_num = perm_num,
                              unif_round_unit = unif_round_unit,
                              weight = weight,
                              shrink = shrink
                            )[[1]]
                          }, error = function(e) {NA}))
                        }}, .progress = "text")
  rownames(op_pval) = gene_id
  op_pval
}
#' This function calculates the BSDE Barycenter p-value of a given gene.
#'
#' @param count_per_gene Gene count. Vector of length n (n is number of cells).
#' @param meta_individual Label of individuals (length n).
#' @param meta_phenotype 0/1 vector of length n (0=control, 1=case).
#' @param perm_num Number of random permutations for computing p-value.
#' @param unif_round_unit Bin width for empirical histogram (default: 0.2).
#' @param weight Weight vector (non-negative, length=number of subjects) for subjects.
#'   Can be used for inverse probability weighting to control for confounding.
#'   Default: 1 for equal weights.
#' @param shrink If \code{TRUE}, the range with zero count will be removed to
#'   cope with sparsity and speed up the computation. Use \code{TRUE} if
#'   extremely high expressions are present. Default: \code{FALSE}.
#' @return a list contains:
#' \itemize{
#'   \item pval: p-value
#'   \item case_bc_ob: density of the aggregated Barycenter distribution of the cases
#'   \item ctrl_bc_ob: density of the aggregated Barycenter distribution of the controls
#' }
#' @export
cal_w2_pval <- function(count_per_gene, meta_individual, meta_phenotype,
                        perm_num=200, unif_round_unit=0.2,
                        weight=1, shrink=FALSE){
  cur_individual=unique(meta_individual)
  phenotype=meta_phenotype[match(cur_individual,meta_individual)]

  n=length(cur_individual)

  #count_per_gene=round(count_per_gene,unif_round_unit)
  cur_range=seq(min(count_per_gene),(max(count_per_gene)+unif_round_unit),by=unif_round_unit) #cur range +1
  unif_round_unit_digit=min(decimalplaces(unif_round_unit),5)
  cur_range=floor(cur_range*10^unif_round_unit_digit)/10^unif_round_unit_digit
  cur_range=c(cur_range, (max(cur_range)+cur_range[2]-cur_range[1])) #cur range expand +2
  len=length(cur_range)-1
  cur_burden=matrix((rep(cur_range[1:len],times=len)-rep(cur_range[1:len],each=len))^2,len,len)
  #cur_burden=cur_burden/(10^unif_round_unit)
  cur_distr=matrix(0,nrow=len,ncol=n) #row for ranges col for individuals
  #background_image=marix(0,nrow=len,ncol=(len-1))
  colnames(cur_distr)=cur_individual
  rownames(cur_distr)=cur_range[1:len]

  for(i_n in 1:n){
    cur_ind=cur_individual[i_n]
    cur_ind_count=count_per_gene[meta_individual==cur_ind]
    cur_freq=graphics::hist(cur_ind_count,breaks=cur_range,plot=FALSE)$counts
    cur_distr[,i_n]=cur_freq/sum(cur_freq)
  }

  if(shrink){ #however, typically, we don't shrink, since it may just by chance that there are no numbers between
    #remove zero counts, to fasten the calculation.
    shrink_index=which(apply(cur_distr,1,sum)>0)
    cur_burden=cur_burden[shrink_index,shrink_index]
    cur_distr=cur_distr[shrink_index,]
    cur_range=cur_range[shrink_index]
    len=length(cur_range)-1
  }

  `%dorng%` <- doRNG::`%dorng%`
  `%dopar%` <- foreach::`%dopar%`
  w2_res=foreach::foreach(ip=0:perm_num) %dorng% {
  #for(ip in 0:10) {
    #print(ip)
    if(ip>0){
      cur_phenotype=phenotype[sample.int(n,n)]
    }
    if(ip==0){
      cur_phenotype=phenotype
    }
    case_distr=cur_distr[,cur_phenotype==1]
    ctrl_distr=cur_distr[,cur_phenotype==0]

    w_case=1
    w_ctrl=1
    if(length(weight)!=1){
      stopifnot("weight length must match # of subjects"=length(weight)==length(cur_phenotype))
      w_case=weight[cur_phenotype==1]
      w_ctrl=weight[cur_phenotype==0]
      w_case =  w_case / sum(w_case)
      w_ctrl =  w_ctrl / sum(w_ctrl)
    }

    py <- reticulate::import_main()
    py$case_wass=NULL
    py$ctrl_wass=NULL
    py$case_distr=reticulate::r_to_py(case_distr)
    py$ctrl_distr=reticulate::r_to_py(ctrl_distr)
    py$w_case=reticulate::r_to_py(w_case)
    py$w_ctrl=reticulate::r_to_py(w_ctrl)

    reticulate::py_run_string("case_wass = cal_bary_wass(case_distr,w=w_case)")
    reticulate::py_run_string("ctrl_wass = cal_bary_wass(ctrl_distr,w=w_ctrl)")
    case_bc=py$case_wass
    ctrl_bc=py$ctrl_wass
    names(case_bc)=cur_range[1:len]
    names(ctrl_bc)=cur_range[1:len]
    #calculate w2

    w2=Barycenter::Greenkhorn(as.matrix(case_bc),(as.matrix(ctrl_bc)),costm=abs(cur_burden), lambda=0.01)$Distance
    #w2
    list(w2,case_bc,ctrl_bc)
  }
  #w2_ob=w2_res[[1]]
  #w2_perm=sapply(w2_res,unlist)[-1]
  w2_ob=w2_res[[1]][[1]]
  w2_perm=sapply(w2_res[-1],function(x)unlist(x[[1]]))
  case_bc_ob=w2_res[[1]][[2]]
  ctrl_bc_ob=w2_res[[1]][[3]]

  pval=mean(w2_perm>=w2_ob,na.rm=TRUE)
  return(list(pval,case_bc_ob,ctrl_bc_ob))
}

#' Calculate p-value for a given gene based on bulk RNAseq.
#' The expressions are represented as one numerical value per gene, per individual.
#'
#' @param cur_ind_count vector of gene counts (length = number of individuals).
#' @param cur_individual individual annotations (length = number of individuals).
#' @param phenotype 0/1 individual labels (length = number of individuals, 0=control, 1=case).
#' @param perm_num number of permutations for computing p-value (integer).
#' @param unif_round_unit Bin width for empirical histogram (default: 0.2).
#' @return a list contains:
#' \itemize{
#' \item pval: p-value
#' \item case_distr: density of the aggregated Barycenter distribution of the cases
#' \item ctrl_distr: density of the aggregated Barycenter distribution of the controls
#' }
#' @export
#' @examples
#' \dontrun{
#' count_per_gene=c(rpois(6,10),c(rpois(6,4)))
#' meta_individual=paste0("ind",rep(1:12,each=1))
#' meta_phenotype=c(rep(1,6),rep(0,6))
#' cal_w2_bulk_pval(count_per_gene,meta_individual,meta_phenotype)
#' }
#'
cal_w2_bulk_pval=function(cur_ind_count,cur_individual,phenotype,
                          perm_num=500,unif_round_unit=0.2){
  n=length(cur_individual)
  cur_range=seq(min(cur_ind_count),(max(cur_ind_count)+unif_round_unit),by=unif_round_unit) #cur range +1
  unif_round_unit_digit=min(decimalplaces(unif_round_unit),5)
  cur_range=floor(cur_range*10^unif_round_unit_digit)/10^unif_round_unit_digit
  cur_range=c(cur_range, (max(cur_range)+cur_range[2]-cur_range[1])) #cur range expand +2
  len=length(cur_range)-1
  cur_burden=matrix((rep(cur_range[1:len],times=len)-rep(cur_range[1:len],each=len))^2,len,len)

  `%dorng%` <- doRNG::`%dorng%`
  `%dopar%` <- foreach::`%dopar%`
  w2_res=foreach::foreach(ip=0:perm_num) %dorng% {
    if(ip>0){
      cur_phenotype=phenotype[sample.int(n,n)]
    }
    if(ip==0){
      cur_phenotype=phenotype
    }
    case_distr=graphics::hist(cur_ind_count[cur_phenotype==1],breaks=cur_range,plot=FALSE)$counts
    ctrl_distr=graphics::hist(cur_ind_count[cur_phenotype==0],breaks=cur_range,plot=FALSE)$counts
    names(case_distr)=cur_range[1:len]
    names(ctrl_distr)=cur_range[1:len]
    w2=Barycenter::Greenkhorn(as.matrix(case_distr),(as.matrix(ctrl_distr)),costm=abs(cur_burden), lambda=0.1)$Distance
    list(w2,case_distr,ctrl_distr)
  }
  w2_ob=w2_res[[1]][[1]]
  w2_perm=sapply(w2_res[-1],function(x)unlist(x[[1]]))
  case_distr=w2_res[[1]][[2]]
  ctrl_distr=w2_res[[1]][[3]]

  pval=mean(w2_perm>=w2_ob,na.rm=TRUE)
  return(list(pval,case_distr,ctrl_distr))
}




inst/op_functions.py
import numpy as np
import ot
import ot.plot

def cal_bary_wass( x, M=None,w=1, reg=2e-2 ): 
  A = np.vstack(x)                    
  n_distributions = A.shape[1]  
  if (M is None):
    M = ot.utils.dist0(A.shape[0])                      
    M /= M.max() 
  if w==1:
    alpha = w/n_distributions
    weights = np.repeat(alpha, n_distributions)
  if w!=1:
    weights = np.array(w)
    weights = weights / sum(weights) #require sum(weights)=1
  
  bary_wass = ot.bregman.barycenter(A, M, reg, weights)
  if(bary_wass[0]!=bary_wass[0] or max(bary_wass)<np.quantile(A, 0.3)):
    reg = 2e-3
    bary_wass = ot.bregman.barycenter(A, M, reg, weights)
  if(bary_wass[0]!=bary_wass[0] or max(bary_wass)<np.quantile(A, 0.3)):
    reg = 2e-4
    bary_wass = ot.bregman.barycenter(A, M, reg, weights)
  return bary_wass


simulation/simulation.R
# This is the example of the simulation experiment with given parameters in the paper.

##################################################################
#
#                    Part I simulation
#
##################################################################

library("abind")
library("emdbook")
library("MASS")
library("DESeq2")
library("moments")
library("MAST")
library("lme4")
library("BSDE")

#senarios
nMean = 3 # of genes shows mean differences
nVar = 3  # of genes shows variance differences
nMult = 3 # of genes shows multimodality differences
nDP = 3   # of genes shows different proportion for mixture models
nGeneBlank = 18
nTotal = nMean + nVar + nMult + nDP + nGeneBlank
ncase = 20   # of case subjects
nctrl = 20   # of ctrl subjects
ncell = 100  # of cells per individual
nall = ncase + nctrl

#simulation of size factors
r_mean = 1.2
r_var = 1.2
r_mult = 0.6
r_dp = 0.2


i_mean = 1:nMean
i_var = (nMean + 1):(nMean + nVar)
i_mult = (nMean + nVar + 1):(nMean + nVar + nMult)
i_dp = (nMean + nVar + nMult + 1):(nMean + nVar + nMult + nDP)
i_blank = (nMean + nVar + nMult + nDP + 1):nTotal

#randomlized settings, switching cases and control samples to make sure the library sizes the same
case_modify_flag = rbinom(nTotal, 1, 0.5)
i_case_modify = which(case_modify_flag == 1)

#first case, then ctrl
i_case = 1:(ncase * ncell)
i_ctrl = (ncase * ncell + 1):(nall * ncell)

# ############ Simulation Data Preparation ###############################

file_tag = paste0(r_mean, "_", r_var, "_", r_mult, "_", r_dp)
dir = "../data/"
setwd(dir)

RIN_adj = TRUE

#readin data
t_mean = readRDS(paste0(dir, "autism/dca_PFCL23_100x10perc/mean.rds"))
t_disp = readRDS(paste0(dir, "autism/dca_PFCL23_100x10perc/dispersion.rds"))
t_drop = readRDS(paste0(dir, "autism/dca_PFCL23_100x10perc/dropout.rds"))
t_meta = readRDS(paste0(dir, "autism/meta_PFCL23_100x10perc.rds"))

dim(t_mean)
t_mean[1:2, 1:5]

dim(t_disp)
t_disp[1:2, 1:5]

dim(t_drop)
t_drop[1:2, 1:5]

table(t_meta$individual)
t_meta$RIN = t_meta$RNA.Integrity.Number

###########simulate basic parameters of ZINB model################
res = simu_base_param(t_mean, t_disp, t_drop, t_meta, nTotal, nall, RIN_adj =
                        RIN_adj)
sample_ctrl = res$sample_ctrl
RIN.simu = res$RIN.simu

random_idx_gene = sample.int(nTotal)
random_idx_sam  = sample.int(nall)

sample_ctrl = sample_ctrl[random_idx_gene, random_idx_sam,]

##########  calculate parameters for the case data #################

# sample gene index for genes differential expressed by mean or variance.
special_index = sample.int(nTotal, (nMean + nVar + nMult + nDP))
mean_index    = as.numeric(special_index[i_mean])
var_index     = as.numeric(special_index[i_var])
mult_index     = as.numeric(special_index[i_mult])
dp_index     = as.numeric(special_index[i_dp])
# label and save the DE index information.
de.mean = rep(0, nTotal)
de.var  = rep(0, nTotal)
de.mult  = rep(0, nTotal)
de.dp  = rep(0, nTotal)
de.mean[mean_index] = 1
de.var[var_index]   = 1
de.mult[mult_index]   = 1
de.dp[dp_index]   = 1
# To make sure all parameters are non-negative, we do some transformation
r_mean2 = r_mean
r_var2  = r_var

if (r_mean < 1) {
  r_mean2 = 1 / r_mean
}

if (r_var < 1) {
  r_var2 = 1 / r_var
}

# modify parameters for cases mean and var
# now r_m (r_mean2) is smaller than 1. We need to modify gene expression in
# x proportion of genes by fold r_m and (1-x) proportion of genes by
# fold of 1/r_m, so that x(1 - r_m) = (1-x)(1/r_m - 1)
# x (1 - r_m + 1/r_m -1) = (1/r_m -1) => x = 1/(1 + r_m)


sample_param_case = abind::abind(sample_ctrl[, 1:ncase,], sample_ctrl[, 1:ncase,], along =
                                   4)#gene x ind x (mean,overdispersion, droupout) x #of mixture models(2)

sample_param_ctrl = abind::abind(sample_ctrl[, (ncase + 1):nall,], sample_ctrl[, (ncase +
                                                                                    1):nall,], along = 4)


for (i in mean_index) {
  #distr1==distr2, case!=ctrl
  for (j in 1:ncase) {
    x = sample_param_case[i, j, , 1]
    sample_param_case[i, j, 1:2, 1] = calc_zinb_param(
      mu = x[1],
      theta = x[2],
      drop = x[3],
      r_m = r_mean2,
      r_v = 1
    )
    sample_param_case[i, j, 1:2, 2] = sample_param_case[i, j, 1:2, 1]
  }
}
for (i in var_index) {
  #distr1==distr2, case!=ctrl
  for (j in 1:ncase) {
    x = sample_param_case[i, j, , 1]
    sample_param_case[i, j, 1:2, 1] = calc_zinb_param(
      mu = x[1],
      theta = x[2],
      drop = x[3],
      r_m = 1,
      r_v = r_var2
    )
    sample_param_case[i, j, 1:2, 2] = sample_param_case[i, j, 1:2, 1]
  }
}

for (i in mult_index) {
  for (j in 1:ncase) {
    #modify mean
    sample_param_case[i, j, 1, 1] = sample_param_case[i, j, 1, 1] + sample_param_case[i, j, 1, 1] *
      (r_mult)
    sample_param_case[i, j, 1, 2] = sample_param_case[i, j, 1, 2] - sample_param_case[i, j, 1, 2] *
      (r_mult)
    #modify disp
    sample_param_case[i, j, 2, 1] = sample_param_case[i, j, 2, 1] * 2
    sample_param_case[i, j, 2, 2] = sample_param_case[i, j, 2, 1]
  }
  for (j in 1:nctrl) {
    sample_param_ctrl[i, j, 2, 1] = calc_multimod_param(size = (sample_param_ctrl[i, j, 2, 1] *
                                                                  2), t = r_mult)
    sample_param_ctrl[i, j, 2, 2] = sample_param_ctrl[i, j, 2, 1]
  }
}

for (i in dp_index) {
  for (j in 1:ncase) {
    x = sample_param_case[i, j, , 1]
    sample_param_case[i, j, 1:2, 1] = calc_zinb_param(
      mu = x[1],
      theta = x[2],
      drop = x[3],
      r_m = 4,
      r_v = 1
    ) #here we set r_m=4
  }
  for (j in 1:nctrl) {
    x = sample_param_ctrl[i, j, , 2]
    sample_param_ctrl[i, j, 1:2, 2] = calc_zinb_param(
      mu = x[1],
      theta = x[2],
      drop = x[3],
      r_m = 4,
      r_v = 1
    ) #here we set r_m=4
  }
}

temp = NA
temp = sample_param_case[i_case_modify, , ,]
sample_param_case[i_case_modify, , ,] = sample_param_ctrl[i_case_modify, , ,]
sample_param_ctrl[i_case_modify, , ,] = temp



# simulate scRNAseq based on zinb parameters of cases and controls:
sim_matrix = matrix(nrow = nTotal, ncol = nall * ncell)
sim_param = array(dim = c(nTotal, nall * ncell, 3))
date()
for (i in 1:nall) {
  idx_i = ((i - 1) * ncell + 1):(i * ncell)
  for (k in 1:ncell) {
    cell_distr_flag = rbinom(1, 1, 0.5) + 1 #indicate the minor prop for simulation
    sub_sample_param_case = sample_param_case[, , , cell_distr_flag] #equally separated for mult as design
    sub_sample_param_ctrl = sample_param_ctrl[, , , cell_distr_flag] #equally separated for mult as design
    cell_distr_flag_dp_add = (cell_distr_flag - 1) * rbinom(1, 1, prob =
                                                              (r_dp * 2)) + 1
    sub_sample_param_case[which(de.dp == 1), ,] = sample_param_case[which(de.dp ==
                                                                            1), , , cell_distr_flag_dp_add] #equally separated for dp as design
    sub_sample_param_ctrl[which(de.dp == 1), ,] = sample_param_ctrl[which(de.dp ==
                                                                            1), , , cell_distr_flag_dp_add] #equally separated for dp as design

    if (i > ncase) {
      mean_i = sub_sample_param_ctrl[, (i - ncase), 1]
      disp_i = sub_sample_param_ctrl[, (i - ncase), 2]
      drop_i = sub_sample_param_ctrl[, (i - ncase), 3]
      sample_mean_sd_i = sub_sample_param_ctrl[, (i - ncase), 4]
    } else{
      mean_i = sub_sample_param_case[, i, 1]
      disp_i = sub_sample_param_case[, i, 2]
      drop_i = sub_sample_param_case[, i, 3]
      sample_mean_sd_i = sub_sample_param_case[, i, 4]
    }

    sample_mean_k = exp(rnorm(nTotal, log(mean_i), sample_mean_sd_i))
    for (ig in 1:nTotal) {
      cur_count = emdbook::rzinbinom(1, sample_mean_k[ig], disp_i[ig], drop_i[ig])
      if (is.na(cur_count)) {
        print(c(i, ik, ig, sample_mean_k[ig], disp_i[ig], drop_i[ig]))
        while (is.na(cur_count)) {
          cur_count = emdbook::rzinbinom(1, sample_mean_k[ig], disp_i[ig], drop_i[ig])
          if (!is.na(cur_count)) {
            print(c(i, ik, ig, cur_count))
          }
        }
      }
      sim_matrix[ig, idx_i[k]] = cur_count
      sim_param[ig, idx_i[k],] = c(sample_mean_k[ig], disp_i[ig], drop_i[ig])
    }
  }
  #print(i)
}
date()

dim(sim_matrix)
sim_matrix[1:4, 1:4]

table(c(sim_matrix) == 0)
table(c(sim_matrix) == 0) / (nrow(sim_matrix) * ncol(sim_matrix))



# scatter plot 2:check simulation
par(mfrow = c(2, 5))

for (ig in which(de.mean + de.var + de.mult + de.dp == 0)[1:2]) {
  for (i in c(1:5, (nall - 4):nall)) {
    idx_i = ((i - 1) * ncell + 1):(i * ncell)
    if (i <= ncase) {
      cur_col = "red"
      cur_label = paste0("non-DE gene ", ig, ", case ")
    }
    if (i >= ncase) {
      cur_col = "blue"
      cur_label = paste0("non-DE gene ", ig, ", ctrl ")
    }
    hist(sim_matrix[ig, idx_i],
         col = cur_col,
         main = cur_label,
         xlab = "count")
  }
}

for (ig in which(de.mean == 1)[1:2]) {
  for (i in c(1:5, (nall - 4):nall)) {
    idx_i = ((i - 1) * ncell + 1):(i * ncell)
    if (i <= ncase) {
      cur_col = "red"
      cur_label = paste0("Mean-DE gene ", ig, ", case ")
    }
    if (i >= ncase) {
      cur_col = "blue"
      cur_label = paste0("Mean-DE gene ", ig, ", ctrl ")
    }
    hist(sim_matrix[ig, idx_i],
         col = cur_col,
         main = cur_label,
         xlab = "count")
  }
}

for (ig in which(de.var == 1)[1:2]) {
  for (i in c(1:5, (nall - 4):nall)) {
    idx_i = ((i - 1) * ncell + 1):(i * ncell)
    if (i <= ncase) {
      cur_col = "red"
      cur_label = paste0("Var-DE gene ", ig, ", case ")
    }
    if (i >= ncase) {
      cur_col = "blue"
      cur_label = paste0("Var-DE gene ", ig, ", ctrl ")
    }
    hist(sim_matrix[ig, idx_i],
         col = cur_col,
         main = cur_label,
         xlab = "count")
  }
}

for (ig in which(de.mult == 1)[1:2]) {
  for (i in c(1:5, (nall - 4):nall)) {
    idx_i = ((i - 1) * ncell + 1):(i * ncell)
    if (i <= ncase) {
      cur_col = "red"
      cur_label = paste0("Mult-DE gene ", ig, ", case ")
    }
    if (i >= ncase) {
      cur_col = "blue"
      cur_label = paste0("Mult-DE gene ", ig, ", ctrl ")
    }
    hist(sim_matrix[ig, idx_i],
         col = cur_col,
         main = cur_label,
         xlab = "count")
  }
}

for (ig in which(de.dp == 1)[1:2]) {
  for (i in c(1:5, (nall - 4):nall)) {
    idx_i = ((i - 1) * ncell + 1):(i * ncell)
    if (i <= ncase) {
      cur_col = "red"
      cur_label = paste0("DP-DE gene ", ig, ", case ")
    }
    if (i >= ncase) {
      cur_col = "blue"
      cur_label = paste0("DP-DE gene ", ig, ", ctrl ")
    }
    hist(sim_matrix[ig, idx_i],
         col = cur_col,
         main = cur_label,
         xlab = "count")
  }
}


####################### Meta information collection #################

#the phenotype and individual information of simulated samples.
phenotype = c(rep(0, nctrl * ncell), rep(1, ncase * ncell))
individual = paste0("ind", c(rep(1:nall, each = ncell)))

#Count info for matrix
cell_id = paste0("cell", 1:(nall * ncell))
gene_id = paste0("gene", 1:nTotal)

rownames(sim_matrix) = gene_id
colnames(sim_matrix) = cell_id

dimnames(sim_param) = list(gene_id, cell_id, c("mean", "overdisp", "dropout"))

#Cell info for meta
cellsum = apply(sim_matrix, 2, sum)
genesum = apply(sim_matrix, 1, sum)
CDR  = apply(sim_matrix > 0, 2, sum) / nrow(sim_matrix)
meta = data.frame(cell_id,
                  individual,
                  phenotype,
                  cellsum,
                  CDR,
                  stringsAsFactors = FALSE)

meta$RIN = rep(RIN.simu, each = ncell) #if RNA_adj==FALSE in simu_base_param, meta$RIN=NA

dim(meta)
meta[1:2,]

######### Some basic stat & Preparation for Bulk RNAseq analysis ############
library(DESeq2)
# individual level info
cur_individual = unique(meta$individual)
cell_num = matrix(ncol = 1, nrow = length(cur_individual))
rownames(cell_num) = cur_individual
colnames(cell_num) = "cell_num"

read_depth = matrix(ncol = 1, nrow = length(cur_individual))
rownames(read_depth) = cur_individual
colnames(read_depth) = "read_depth"

phenotype_ind = matrix(ncol = 1, nrow = length(cur_individual))
rownames(phenotype_ind) = cur_individual
colnames(phenotype_ind) = "phenotype"

zero_rate_ind = matrix(nrow = nrow(sim_matrix),
                       ncol = length(cur_individual))
rownames(zero_rate_ind) = rownames(sim_matrix)
colnames(zero_rate_ind) = cur_individual

sim_matrix_bulk = matrix(nrow = nrow(sim_matrix),
                         ncol = length(cur_individual))
rownames(sim_matrix_bulk) = rownames(sim_matrix)
colnames(sim_matrix_bulk) = cur_individual

for (i_ind in 1:length(cur_individual)) {
  cur_ind = cur_individual[i_ind]
  #fit org
  cur_ind_m = sim_matrix[, meta$individual == cur_ind]
  cell_num[i_ind]   = ncol(cur_ind_m)
  read_depth[i_ind] = sum(cur_ind_m, na.rm = TRUE) / cell_num[i_ind] * 1000
  phenotype_ind[i_ind] = meta$phenotype[meta$individual == cur_ind][1]

  zero_rate_ind[, i_ind] = rowSums(cur_ind_m == 0, na.rm = TRUE) / cell_num[i_ind]
  sim_matrix_bulk[, i_ind] = rowSums(cur_ind_m, na.rm = TRUE)
}

tapply(read_depth, phenotype_ind, summary)

########################Other cell level info################################
library("MAST")
library("lme4")
cell_readdepth = apply(sim_matrix, 2, sum)
cell_count_quantile = apply(sim_matrix, 2, quantile)

#almost no need to adjust library size.
# saveRDS(de.mean,
#         paste0(dir, "simulated/sim_de.mean_", file_tag, ".rds"))
# saveRDS(de.var, paste0(dir, "simulated/sim_de.var_", file_tag, ".rds"))
# saveRDS(de.mult,
#         paste0(dir, "simulated/sim_de.mult_", file_tag, ".rds"))
# saveRDS(de.dp, paste0(dir, "simulated/sim_de.dp_", file_tag, ".rds"))
#
# saveRDS(read_depth,
#         paste0(dir, "simulated/sim_ind_readdepth_", file_tag, ".rds"))
# saveRDS(phenotype_ind,
#         paste0(dir, "simulated/sim_ind_phenotye_", file_tag, ".rds"))
# saveRDS(zero_rate_ind,
#         paste0(dir, "simulated/sim_ind_zero_rate_", file_tag, ".rds"))
#
# saveRDS(sim_param,
#         paste0(dir, "simulated/sim_param_", file_tag, ".rds"))
# saveRDS(sim_matrix,
#         paste0(dir, "simulated/sim_matrix_", file_tag, ".rds"))
# saveRDS(meta, paste0(dir, "simulated/sim_meta_", file_tag, ".rds"))
# saveRDS(sim_matrix_bulk,
#         paste0(dir, "simulated/sim_matrix_bulk_", file_tag, ".rds"))
#
# saveRDS(cell_readdepth,
#         paste0(dir, "simulated/sim_cell_readdepth_", file_tag, ".rds"))
# saveRDS(
#   cell_count_quantile,
#   paste0(dir, "simulated/sim_cell_count_quantile_", file_tag, ".rds")
# )
#
# write.csv(sim_matrix,
#           paste0(dir, "simulated/sim_matrix_", file_tag, ".csv"))


##################################################################
#
#                    Part II simulation
#
##################################################################

###DESeq2######

# We calculate bulk information by summing up raw counts of
# all cells(of certain cluster) of an individual within a genes

cur_info = meta[, c("individual", "phenotype")]
cur_info = unique(cur_info)
rownames(cur_info) = cur_info$individual
cur_info$phenotype = as.factor(cur_info$phenotype)

# object construction
dds = DESeqDataSetFromMatrix(countData = sim_matrix_bulk,
                             colData = cur_info,
                             design = ~ phenotype)

# observed pvalue calculation
dds = DESeq(dds)
deseq_pval = results(dds)$pvalue


#####MAST######

#we take the method of current usage of MAST,
#this code care with the results of simulation data, and calculate the MAST


# the input of MAST analysis can be many format including matrix and
# SingleCellAssay.

# input:
# (1)the log-transformed expression count matrix,
#     with each column represents the cells and each rows represents the genes.
# (2)the meta data, including cell information and individual information.

# we get the p-values based on the Hurdle model ("H" model)

sim_matrix_log = log2(1 + sim_matrix) #log transformed data

dim(sim_matrix_log)
sim_matrix_log[1:2, 1:5]

cell_id = colnames(sim_matrix_log)   #get the cell id from the data
gene_id = rownames(sim_matrix_log)   #get the gene id from the data

fData = data.frame(primerid = gene_id)
cData = data.frame(wellKey = cell_id)

diagnosis = as.character(meta$phenotype) #

diagnosis2 = matrix("Control", ncol = 1, nrow = length(diagnosis))
diagnosis2[which(diagnosis == "1")] = "Case"
diagnosis = as.factor(diagnosis2)

sca = MAST::FromMatrix(sim_matrix_log, cData, fData)
colData(sca)$cngeneson = as.numeric(meta$CDR)
colData(sca)$diagnosis = as.factor(diagnosis)
colData(sca)$ind = as.factor(meta$individual)

colData(sca)

print(date())
print(gc())
b1 = tryCatch(
  MAST::zlm(
    formula = ~ diagnosis + (1 |
                               ind),
    sca = sca,
    method = 'glmer',
    ebayes = FALSE,
    parallel = TRUE
  ),
  error = function(e) {
    NA
  }
)
print(date())
print(gc())

lrt1 = tryCatch(
  MAST::lrTest(b1, "diagnosis"),
  error = function(e) {
    NA
  }
)
MAST_pval1 = tryCatch(
  apply(lrt1, 1, function(x) {
    x[3, 3]
  }),
  error = function(e) {
    NA
  }
)
print(date())
print(gc())

####### BSDE ##########

perm_num = 500
lambda_fold = 1.5
unif_round_unit = 0.05
library("reticulate")
library("doRNG")
library("doParallel")
library("ggplot2")
library("emdbook")
library("Rcpp")
library("Barycenter")
library("BSDE")

op_pval = matrix(ncol = 1, nrow = nrow(sim_matrix_log))


print(date())
print(gc())

for (i_g in 1:nrow(sim_matrix_log)) {
  cur_sim = sim_matrix_log[i_g, ]
  cur_ind = meta$individual
  cur_pheno = phenotype

  op_pval[i_g] = tryCatch({
    cal_w2_pval(
      count_per_gene = cur_sim,
      meta_individual = cur_ind,
      meta_phenotype = cur_pheno,
      perm_num = perm_num,
      unif_round_unit = 0.5
    )[[1]]
  }, error = function(e) {
    NA
  })
}
print(date())
print(gc())
rownames(op_pval) = gene_id

##################################################################
#
#              Part III simulation result analysis
#
##################################################################
